### ISSUE TITLE
bug: Borgmatic yields deprecation warning

### ISSUE THREAD

**Issue by randomn4me** [nix-community/home-manager#4580]
### Are you following the right branch?

- [X] My Nixpkgs and Home Manager versions are in sync

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Issue description

Running the current version of borgmatic (Version 1.8.1) warns the user, that "The repositories option now expects a list of key/value pairs. Lists of strings for this option are deprecated and support will be removed from a future release."

In the current borgmatic module the repositories are just defined as `types.listOf types.str` [[1]](https://github.com/nix-community/home-manager/blob/master/modules/programs/borgmatic.nix#L59). I already created a [commit](https://github.com/randomn4me/home-manager/commit/a1a0ff915c2f0e543992c6a703970b28ca89a89c), which solves this issue, but it is not backwards compatible and yields no warning, as suggested [in the manual](https://nix-community.github.io/home-manager/index.html#sec-guidelines-back-compat). Is there anything else that needs to be done or can I just create a Pull Request.

Cheers,
r4ndom

### Maintainer CC

@DamienCassou

### System information

```markdown
- system: `"x86_64-linux"`
- host os: `Linux 6.5.7, NixOS, 23.11 (Tapir), 23.11.20231011.5e4c2ad`
- multi-user?: `yes`
- sandbox: `yes`
- version: `nix-env (Nix) 2.17.0`
- nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```
```

**Comment by rycee** [comment]
I think the software changing its configuration format would be a valid reason to change the options in a backwards-incompatible way.

A possibility would be to change the type from `types.listOf repositoryOption` to `types.listOf (types.either str repositoryOption)` and then issue a warning if a string value is given. Then, after some suitable time, one could remove the `str` alternative.

**Comment by DamienCassou** [comment]
I like @rycee's suggestion to allow both formats make sense because [borgmatic says that labels are optional](https://torsion.org/borgmatic/docs/reference/configuration/):

> A required list of local or remote repositories with paths and optional labels

I would go as far as *not* issuing a warning if a string is given: if labels are optional, a list of paths makes perfect sense to me.

**Comment by rycee** [comment]
Aha, in that case I would suggest not doing a warning and instead make the label have the type `types.nullOr types.str` (with `null` being the default) and include
``` nix
apply = val: if builtins.isString val then { path = val; } else val;
```
in the option.

**Comment by randomn4me** [comment]
I now got (something)[https://github.com/randomn4me/home-manager/commit/53bdb2be4e998448f68c757660668bf835087a6a], that works when

1. there is only strings in the repos list
2. the repositoryOption is filled with label and path

If the label is not given by the user borgmatic complains because the final configuration file looks like this

```
  "repositories": [
    {
      "label": null,
      "path": "path/to/repo"
    }
  ],
```

I don't know, why the label is still there after `removeNullValues`.

**Comment by DamienCassou** [comment]
@randomn4me do you mind opening a PR?

**Comment by randomn4me** [comment]
@DamienCassou Done.

**Comment by randomn4me** [comment]
Closed by now merged Pull Request #4588
